simple/app.go

```diff
@@ -0,0 +1,334 @@
+	AppManifest app.Manifest
+	NewAppFunc  func(config app.Config) (app.App, error)
+}
+
+// Manifest returns the AppManifest in the AppProvider
+func (a *AppProvider) Manifest() app.Manifest {
+	return a.AppManifest
+}
+
+// NewApp calls NewAppFunc and returns the result
+func (a *AppProvider) NewApp(settings app.Config) (app.App, error) {
+	return a.NewAppFunc(settings)
+}
+
+// NewAppProvider is a convenience method for creating a new AppProvider
+func NewAppProvider(manifest app.Manifest, newAppFunc func(cfg app.Config) (app.App, error)) *AppProvider {
+	return &AppProvider{
+		AppManifest: manifest,
+		NewAppFunc:  newAppFunc,
+	}
+}
+
+var (
+	_ app.App = &App{}
+)
+
+// App is a simple, opinionated implementation of resource.App.
+// It must be created with NewApp to be valid.
+type App struct {
+	informerController *operator.InformerController
+	operator           *operator.Operator
+	clientGenerator    resource.ClientGenerator
+	kinds              map[string]AppManagedKind
+	internalKinds      map[string]resource.Kind
+	cfg                AppConfig
+	converters         map[string]Converter
+}
+
+// AppConfig is the configuration used by App
+type AppConfig struct {
+	Name           string
+	KubeConfig     rest.Config
+	InformerConfig AppInformerConfig
+	ManagedKinds   []AppManagedKind
+	Converters     map[schema.GroupKind]Converter
+}
+
+// AppInformerConfig contains configuration for the App's internal operator.InformerController
+type AppInformerConfig struct {
+	ErrorHandler       func(context.Context, error)
+	RetryPolicy        operator.RetryPolicy
+	RetryDequeuePolicy operator.RetryDequeuePolicy
+	FinalizerSupplier  operator.FinalizerSupplier
+}
+
+// AppManagedKind is a Kind and associated functionality used by an App.
+type AppManagedKind struct {
+	// Kind is the resource.Kind being managed. This is equivalent to a kubernetes GroupVersionKind
+	Kind resource.Kind
+	// Reconciler is an optional reconciler to run for this Kind. Only one version of a Kind should have a Reconciler,
+	// otherwise, duplicate events will be received.
+	Reconciler operator.Reconciler
+	// Watcher is an optional Watcher to run for this Kind. Only one version of a Kind should have a Watcher,
+	// otherwise, duplicate events will be received.
+	Watcher operator.ResourceWatcher
+	// Validator is an optional ValidatingAdmissionController for the Kind. It will be run only for validation
+	// of this specific version.
+	Validator resource.ValidatingAdmissionController
+	// Mutator is an optional MutatingAdmissionController for the Kind. It will be run only for mutation
+	// of this specific version.
+	Mutator resource.MutatingAdmissionController
+	// SubresourceRoutes are an optional map of subresource paths to a route handler.
+	// If supported by the runner, calls to these subresources on this particular version will call this handler.
+	SubresourceRoutes map[string]func(ctx context.Context, writer http.ResponseWriter, req *app.SubresourceRequest) error
```

Comment:
I think we want to route based on at least verb + path / subresource name – should we create a custom struct for this?

Reply:
The payload of the request includes the verb, so I was on the fence about whether the handler should be verb-specific or agnostic. I can try introducing verb-specific handlers here and see how it feels ergonomically.

Currently, there are no runners that use custom subresource routes, so it can also be migrated in the future before we introduce them if necessary, but I'd like to get it right in this PR if possible.

Reply:
I think matching should be done in the sdk as much as possible, so let's move to verb + path.